etc/redfishpower-cray-r272z30.dev

```diff
@@ -33,6 +33,8 @@ specification "redfishpower-cray-r272z30" {
 		expect "redfishpower> "
 		send "setcyclepath redfish/v1/Systems/Self/Actions/ComputerSystem.Reset {\"ResetType\":\"ForceRestart\"}\n"
 		expect "redfishpower> "
+		send "settimeout 60\n"
```

Comment:
Hang on, how does this even work?  If the redfish timeout is reached, then the on/off command is not going to return a string that matches what the device script expects, so powerman will sit there until its timeout is reached, and then restat  redfishpower.  Is there any point to having a timeout for on/off commands in redfishpower if this is the case?

It's a different story with status since that can be made to return `unknown` and be matched in the expect state but not assigned as a plug state (leaving it unknown).  So it could have its own timeout that is much lower than the powerman timeout to make sure that returns quickly.

Reply:
> Hang on, how does this even work? If the redfish timeout is reached, then the on/off command is not going to return a string that matches what the device script expects, so powerman will sit there until its timeout is reached, and then restat redfishpower. Is there any point to having a timeout for on/off commands in redfishpower if this is the case?

Lets say the average time it takes to turn on and verify something is on is 70 seconds.  With redfishpower's current 60 second timeout, an "on" will never work.  It'll report "timeout" all the time.  So we need to have a way to increase it so that it can appropriately wait a little longer and report "ok" when the "on" is done.

So I think in this made up example, we want powerman's timeout and redfishpower's timeout to be (lets say) 90 seconds.  powerman's timeout is 90 seconds so it won't give up on the script too quickly.  and redfishpower's increased timeout is to make sure it has the right amount of time to wait.

Reply:
Yeah but why have both timeouts?  There is no way for the redfish timeout to occur and not also trigger the powerman timeout.  At that point, may as well just use the powerman timeout.

Reply:
The problem is the redfishpower timeout is independent of the powerman timeout.

Using the 70 second "on" example earlier.  If powerman timeout is 90 seconds and redfishpower's timeout stays at the default 60 seconds, then "on" will never work.  Redfish will always report "timeout" to powerman.  We want to set redfishpower's timeout to (lets say) 90 seconds, so it has the chance to report "success" to powerman.

Reply:
oh wait, i think I get what you're saying now.  Are you basically saying don't have a timeout in `redfishpower`?  Just have it spin infinitely for "on" or "off" to complete?  And if powerman times out, it'll kill redfishpower?

Reply:
> Yeah, does that seem right?

It does seem wrong to spin infinitely :-)  But I think the idea makes sense.  I don't know how powerman behaves in this particular case.  Will it still say "Command Completed successfully"?  I'll have to test later on.

The timeout is useful for testing when I run redfishpower by hand (i.e. get more debug info than when running under powerman), but perhaps I'll just increase the timeout to like INT_MAX or something, and that'll serve the right purpose here.

In addition, given #76, will redfishpower be able to respond "timeout" to powerman in the future?

Reply:
This idea may be dumb:  what if we made redfishpower into a somewhat usable tool on its own, and then had a powerman mode that is selected by the dev scripts? (`--powerman-helper` or similar)?  Then you could do stuff like fire off a power on or power off command to an arbitrary hostlist and use powerman's fairly stable pm -q to watch it without being stuck behind a power action?

E.g.
```
$ redfishpower on node[1-1024]
$ pm -q
off: node[512-1024]
on: node[1-256]
unknown: node[257-511]
```

Just giving the admins more tools when the nodes are annoying to control with powerman.

Reply:
> It should tell you the command failed and exit with an error code.

Yup got

```
redfishpower-peaks: action timed out waiting for expected response
Command completed with errors
```

> Just had an idea: what if on timeout, redfishpower just exited with an error? That might be a way to fail fast. Experiment > needed.

I don't think this does what we want. (the "XXX ..." below is my debug, ignore that)

```
>pm -T --on hetchy33
send(redfishpower-peaks): 'on phetchy33\n'
send(redfishpower-peaks): 'on phetchy33\n'
send(redfishpower-peaks): 'on phetchy33\n'
recv(redfishpower-peaks): 'XXX PoweringOn\nXXX PoweringOn\nXXX PoweringOn\nXXX On\nphetchy33: ok\nredfishpower> '
Command completed successfully
```

I'm guessing redfishpower exists, powerman restarts, and re-issues on command several times.  It just "happens to work" on the 3rd attempt.

Soo .... for the short term, I think making the redfishpower timeout "infinite" might be a good strategy.  B/c atleast the user will get a slightly superior error.  Lemme ping folks and see what they think.

Reply:
so no big objections I think.  But I think the major thing for me is issue #76.  Reporting what specific nodes are having issues is a major problem right now.  So if we went with the redfishpower "infinite" timeout,  would it only be a minor incremental improvement before a #76 solution?

One concern is, doing a  big power command like

```
$ pm --off foo[1-10000]
```

may simply return "command returned error" everytime instead of "command successful" everytime, b/c inevitably some node will always have a problem.

Sooo ... if we think that #76 will have a solution in the half-near future, I'm not sure if going with "super large timeout" is a good short term win.

Reply:
I'm not sure what we're going to get done in the near term, but I agree #76 should be high on our list.

I got sort of lost here.  Right now does a power command succeed (exit 0) when the command times out underneath?  If so that's an egregious bug IMHO.

Reply:
Yes, my understanding is that "command completed successfully" is always output.

Example me setting redfishpower timeout to 20 seconds, powerman timeout to 60 seconds, on normally takes around 40 seconds to do an "on".

```
$ time pm -T --off hetchy33; time pm -T --on hetchy33
send(redfishpower-peaks): 'off phetchy33\n'
recv(redfishpower-peaks): 'phetchy33: ok\nredfishpower> '
Command completed successfully
0.000u 0.001s 0:07.97 0.0%	0+0k 0+8io 0pf+0w

send(redfishpower-peaks): 'on phetchy33\n'
recv(redfishpower-peaks): 'phetchy33: timeout\nredfishpower> '
Command completed successfully
0.000u 0.001s 0:21.30 0.0%	0+0k 0+0io 0pf+0w
```

I suspect it's due to the way the scriptlet is normally written?

```
	script on_ranged {
		send "on %s\n"
		expect "redfishpower> "
	}
```

b/c "redfishpower > " is eventually seen, it's always "command completed successfully"?  Related to discussion in #79??

Reply:
is this issue covered by #79?  Or do you feel is different issue?

Reply:
I'll go ahead and open a separate issue.  The problem is that the expect script allows failure to be interpreted as success.  It should only expect success and fail otherwise (after timeout, unfortunately, but that's what we have to work with right now).